### PR TITLE
Stream deck dynamic button rework

### PR DESCRIPTION
**Pull request by slyjeff**
Massive rework to use event bus and reduce direct dependencies between classes.

A decicion I wrestled with: the ImageService and DynamicInfoStore are storage areas for information received from the app; rather than the tcp send/receive services updating these via events, I gave them direct knowledge of the classes and had them call directly. The reason is that these really should be the only consumer of these "events"- anyone else subscribing to these could really cause problems, as they are intended to be the masters of the data that they receive. So I decided (but I'm not certain), it made sense for these to not be calls, and to just update the data directly and then, once the data is stored where it is supposed to be, let those classes raise events for others to react to.

I'm open to changing this.

### CODE REVIEW

File: StreamDeckPlugin/Actions/DynamicAction.cs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+                    return Deck.Player1;
+                }
+
+                return _settings.Deck.AsDeck();
+            }
+        }
+
+        public int Index {
+            get {
+                var rows = 4;
```
Comment (davilimap) [inline StreamDeckPlugin/Actions/DynamicAction.cs:47]: I wonder if this is something we want to make customizable?

File: StreamDeckPlugin/Actions/DynamicAction.cs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+                var columns = 16;
+                var device = StreamDeck.Info.Devices.FirstOrDefault(x => x.Id == _deviceId);
+                if (device != null) {
+                    rows = device.Size.Rows;
+                    columns = device.Size.Columns;
+                }
+
+                var buttonsPerPage = rows * columns - 4; //3 because the return to parent, show hand, left, and right buttons take up four slots
```
Comment (davilimap) [inline StreamDeckPlugin/Actions/DynamicAction.cs]: Update comment

File: StreamDeckPlugin/Actions/DynamicAction.cs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+        private void KeyHeldDown(object sender, ElapsedEventArgs e) {
+            lock (_keyUpLock) {
+                if (!_keyIsDown) {
+                    return;
+                }
+                _keyIsDown = false;
+                _keyPressTimer.Enabled = false;
+
+                SendClick(false);
```
Comment (davilimap) [inline StreamDeckPlugin/Actions/DynamicAction.cs]: Nit: For readability (especially in PRs) I'm a fan of using explicit parameters for bool values. So here, it would be `SendClick(isLeftClick: false)`

File: StreamDeckPlugin/Actions/DynamicAction.cs [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        protected override Task OnKeyUp(ActionEventArgs<KeyPayload> args) {
+            lock (_keyUpLock) {
+                if (!_keyIsDown) {
+                    return Task.CompletedTask;
+                }
+                _keyIsDown = false;
+                _keyPressTimer.Enabled = false;
+
+                _settings = args.Payload.GetSettings<ActionWithDeckSettings>();
+                SendClick(true);
```
Comment (davilimap) [inline StreamDeckPlugin/Actions/DynamicAction.cs]: Nit: Same as above.

File: StreamDeckPlugin/Actions/DynamicAction.cs [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+            //setting the card name, just because we want the button to update to show the opration is finished (no longer have the "pressed in" look)
+            SetTitleAsync(TextUtils.WrapTitle(_lastSetTitle));
+        }
+
+        private void GetButtonInfo() {
+            _eventBus.GetButtonInfo(Deck, Index, Mode);
+        }
+
+        private bool DynamicActionMatchesButton(IDynamicActionInfo dynamicActionInfo) {
+            return (dynamicActionInfo.Deck == Deck && dynamicActionInfo.Index == Index && dynamicActionInfo.Mode == Mode);
```
Comment (davilimap) [inline StreamDeckPlugin/Actions/DynamicAction.cs]: Nit: parentheses aren't needed here

File: StreamDeckPlugin/Actions/DynamicAction.cs [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+            //setting the card name, just because we want the button to update to show the opration is finished (no longer have the "pressed in" look)
+            SetTitleAsync(TextUtils.WrapTitle(_lastSetTitle));
+        }
+
+        private void GetButtonInfo() {
+            _eventBus.GetButtonInfo(Deck, Index, Mode);
+        }
+
+        private bool DynamicActionMatchesButton(IDynamicActionInfo dynamicActionInfo) {
```
Comment (davilimap) [inline StreamDeckPlugin/Actions/DynamicAction.cs]: Minor: I feel like this could be an extension method

File: StreamDeckPlugin/Actions/DynamicAction.cs [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        }
+
+        private void UpdateButtonDisplay(IDynamicActionInfo dynamicActionInfo) {
+            _lastSetTitle = dynamicActionInfo.Text;
+            SetTitleAsync(TextUtils.WrapTitle(_lastSetTitle));
+
+            if (_imageService.HasImage(dynamicActionInfo.ImageId)) {
+                SetImageAsync(_imageService.GetImage(dynamicActionInfo));
+            } else {
+                SetImageAsync(string.Empty);
```
Comment (davilimap) [inline StreamDeckPlugin/Actions/DynamicAction.cs]: Minor: consider adding a ClearImage method to use here.

File: StreamDeckPlugin/Program.cs [hunk 8]
```diff
@@ -10,14 +9,31 @@ class Program {
             // optional, but recommended
             //System.Diagnostics.Debugger.Launch();
```
Comment (davilimap) [inline StreamDeckPlugin/Program.cs]: Not part of this PR, but this should probably get uncommented or removed.

File: StreamDeckPlugin/Events/RegisterForUpdates.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
+﻿using StreamDeckPlugin.Services;
+using System;
+
+namespace StreamDeckPlugin.Events {
+    public class RegisterForUpdates : IEvent {
```
Comment (davilimap) [inline StreamDeckPlugin/Events/RegisterForUpdates.cs]: This event seems unclear to me.  Does the event represent registering for updates or requesting updates?

File: StreamDeckPlugin/Utils/ImageUtils.cs [hunk 10]
```diff
@@ -37,20 +33,7 @@ public static class ImageUtils {
...
-            var converter = new ImageConverter();
-            var converted = (byte[])converter.ConvertTo(bitmap, typeof(byte[]));
-            var imageString = Convert.ToBase64String(converted);
-
-            return "data:image/png;base64," + imageString;
-        }
-
-
-        private static void DrawSelected(Bitmap bitmap) {
+        public static void DrawSelected(Bitmap bitmap) {
```
Comment (davilimap) [inline StreamDeckPlugin/Utils/ImageUtils.cs]: Minor: Not part of this PR, but I think the name here could be clearer. I initially read it as Draw Selected Card

File: StreamDeckPlugin/Utils/RegisterForUpdatesService.cs [hunk 11]
```diff
@@ -1,14 +1,22 @@
...
+using StreamDeckPlugin.Services;
 using System.Timers;
 
 namespace StreamDeckPlugin.Utils {
-    public class RegisterForUpdatesService {
+    public interface IRegisterForUpdatesService {
+        void RegisterForUpdates();
+    }
+
+    public class RegisterForUpdatesService : IRegisterForUpdatesService {
```
Comment (davilimap) [inline StreamDeckPlugin/Utils/RegisterForUpdatesService.cs]: Similarly to previous comment, I think this service can be renamed to something like RequestUpdateService

File: StreamDeckPlugin/Services/SendEventHandler.cs [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+            eventBus.OnGetStatValue(GetStatValue);
+            eventBus.OnChangeStatValue(ChangeStatValue);
+            eventBus.OnGetButtonImage(GetButtonImage);
+        }
+
+        private string SendRequest(Request request) {
+            return SendSocketService.SendRequest(request, TcpInfo.ArkhamOverlayPort);
+        }
+
+        private T SendRequest<T>(Request request) where T : Response {
```
Comment (davilimap) [inline StreamDeckPlugin/Services/SendEventHandler.cs:38]: Doesn't seem like we're using the return on this right now, so maybe we can combine with the above SendRequest method

File: StreamDeckPlugin/Actions/DynamicAction.cs (hunk 1 above)
Comment (slyjeff) [reply StreamDeckPlugin/Actions/DynamicAction.cs:47]: This is ultimately going to change to be smarter by looking to see how many other dynamic buttons exist and then doing the necessary math. Part of this rework was to make this easier to do.

File: StreamDeckPlugin/Actions/DynamicAction.cs (hunk 3 above)
Comment (slyjeff) [reply StreamDeckPlugin/Actions/DynamicAction.cs]: 100% This was an enumeration for that reason, but this is the best idea I think.

File: StreamDeckPlugin/Program.cs (hunk 8 above)
Comment (slyjeff) [reply StreamDeckPlugin/Program.cs]: I need to learn to run in release mode when I don't want to debug

File: StreamDeckPlugin/Actions/DynamicAction.cs (hunk 6 above)
Comment (slyjeff) [reply StreamDeckPlugin/Actions/DynamicAction.cs]: done

File: StreamDeckPlugin/Actions/DynamicAction.cs (hunk 7 above)
Comment (slyjeff) [reply StreamDeckPlugin/Actions/DynamicAction.cs]: done

File: StreamDeckPlugin/Events/RegisterForUpdates.cs (hunk 9 above)
Comment (slyjeff) [reply StreamDeckPlugin/Events/RegisterForUpdates.cs]: This is the code that makes the UI aware that it exists. So it's registering itself so that the UI will send updates to it. Bascially, if this never happens, the streamdeck will never get pushed updates. I have renamed this to "RegisterForUpdatesRequest".

File: StreamDeckPlugin/Utils/ImageUtils.cs (hunk 10 above)
Comment (slyjeff) [reply StreamDeckPlugin/Utils/ImageUtils.cs]: Updated to "DrawSelectedRingOnImage"

File: StreamDeckPlugin/Services/SendEventHandler.cs (hunk 12 above)
Comment (slyjeff) [reply StreamDeckPlugin/Services/SendEventHandler.cs:38]: This allows the calls that don't care about the response to execute without having to put a response type in.

Example: SendRequest(request);

**Review by davilimap — APPROVED** [review]
